### ISSUE TITLE
Exporting layers return a write too long error

### ISSUE THREAD

**Issue by fagiani** [buildpacks/lifecycle#925]
### Summary

I've got an odd error while exporting the `pack build` image:
```
===> EXPORTING
Reading buildpack directory: /layers/heroku_ruby
Reading buildpack directory item: app.toml
Reading buildpack directory item: launch.toml
Reading buildpack directory item: profile
Reading buildpack directory item: profile.toml
Reading buildpack directory item: shim
Reading buildpack directory item: shim.toml
Processing buildpack directory: /layers/heroku_ruby
Processing launch layer: /layers/heroku_ruby/profile
Reusing tarball for layer "heroku/ruby:profile" with SHA: sha256:0e0e120efcfbd6c2a25fbdf3d863b24d356533154398844dbfac2397b9741e5f
Adding layer 'heroku/ruby:profile'
Layer 'heroku/ruby:profile' SHA: sha256:0e0e120efcfbd6c2a25fbdf3d863b24d356533154398844dbfac2397b9741e5f
Reading buildpack directory: /layers/heroku_procfile
Reading buildpack directory item: launch.toml
Processing buildpack directory: /layers/heroku_procfile
ERROR: failed to export: exporting app layers: creating app layers: archive/tar: write too long
ERROR: failed to build: executing lifecycle: failed with status code: 62

[Container] 2022/10/05 12:52:08 Command did not exit successfully dockerd-entrypoint.sh main.sh exit status 1
[Container] 2022/10/05 12:52:08 Phase complete: BUILD State: FAILED_WITH_ABORT
[Container] 2022/10/05 12:52:08 Phase context status code: COMMAND_EXECUTION_ERROR Message: Error while executing command: dockerd-entrypoint.sh main.sh. Reason: exit status 1
```
---

##### Steps

1. Set the following `project.toml`
```
[project]
id = "com.company.project"
name = "MySystem"
version = "1.0"

[build]
builder = "heroku/buildpacks:18"

[[build.buildpacks]]
uri = "heroku/ruby"

[[build.buildpacks]]
uri = "heroku/procfile@1.0.1"
```
2. Set the following `Procfile`
```
web:            bundle exec unicorn -p $PORT -c config/unicorn.rb
clock:          bundle exec rails runner bin/clock
worker:         bundle exec env rake resque:work --trace QUEUE=*
sqs:            bundle exec shoryuken -C config/shoryuken.yml -R
kafka:          bundle exec racecar MyKafkaConsumer
scheduledtasks: lib/tasks/run_task.conf
```
4. Run pack command: `pack build image-name:latest  --tag image-name:1.0 --env-file .env --publish --trust-builder --pull-policy if-not-present --cache-image image-name:cache`

##### Current behavior

Fails with an error when exporting layers.

##### Expected behavior

Finish exporting layers and push to the image repository.

---

##### pack info

```
Pack:
  Version:  0.24.0
  OS/Arch:  linux/amd64

Default Lifecycle Version:  0.13.3

Supported Platform APIs:  0.3, 0.4, 0.5, 0.6, 0.7, 0.8

Config:
  [[trusted-builders]]
    name = "[REDACTED]"
```
##### docker info

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.8.2-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)

Server:
 Containers: 19
  Running: 1
  Paused: 0
  Stopped: 18
 Images: 153
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc version: v1.1.2-0-ga916309
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.15.0-1015-aws
 Operating System: Ubuntu 20.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 3.775GiB
 Name: ip-172-31-79-151
 ID: U5II:ROCG:J2QE:3JGT:ZBP6:ZUM4:SOQE:O2YK:Z6AX:33WF:3I6Y:BHGA
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by natalieparellano** [comment]
@fagiani just circling back on this - were you able to get to the bottom of this?

**Comment by fagiani** [comment]
@natalieparellano unfortunately I did not. The only way I could reproduce the issue was using a trusted buildpack and looks like it follows a different path from a regular one. I'd be still interested in resolving as this issue has happened a couple other random times in specific workloads but i am not sure how to address it ATM.

**Comment by natalieparellano** [comment]
Thanks @fagiani ...I'd like to be helpful here, but sadly I've been unable to reproduce this myself. Working with the sample ruby app (https://github.com/buildpacks/samples/tree/main/apps/ruby-bundler), and running the following command:
```
pack build test-bug --builder heroku/buildpacks:18 --buildpack heroku/ruby --buildpack heroku/procfile@1.0.1 --trust-builder
```
...I don't see the error. Given the error message contains `creating app layers` (which is emitted [here](https://github.com/buildpacks/lifecycle/blob/4c62475a8c58f9026c5e1e4ebea5e90060eabc0d/exporter.go#L315)) I have to guess that the content that is "too long" is in the app directory itself. When this happens, is there any way you can get a read on what is actually in the app?

**Comment by fagiani** [comment]
@natalieparellano As I cannot share the workload where the error happens, I will try to match the ruby sample above with the very settings of the production one to see if the issue happens. If I remember correctly it has something to do with the buildpacks order... (like if procfile gets before ruby or if you add node's one between...). I'll get back to you here as I figure that out. Thanks again!

**Comment by fagiani** [comment]
@natalieparellano I was unable to reproduce the issue with the sample ruby app. I am still able to reproduce that on the one that caused the creation of this very issue. @jjbustamante has [mentioned](https://cloud-native.slack.com/archives/C0331B61A1Y/p1665409037330549?thread_ts=1664975904.790139&channel=C0331B61A1Y&message_ts=1665409037.330549) at our thread in the slack channel [this file](https://github.com/buildpacks/pack/blob/495ef2a4ef38f02996434cd894ef50b2d8dc1ecd/pkg/client/build.go#L374) as the place where it defines which lifecycle version/binary to use and somehow it doesn't matter the code I place debugging flags as they won't run. If I understand correctly, it will kinda download a lifecycle from a trusted source and make sure it is not 'tampered'... Please let me know if you could help me along to be able to tweak it to be able to debug. 

Thanks again!

**Comment by natalieparellano** [comment]
@fagiani if you're bringing your own lifecycle, you can run `pack build --lifecycle-image <custom lifecycle image>` to ensure that your lifecycle gets used for the export phase, or you can just use `pack build --trust-builder` to use the lifecycle that is in your builder already. If you need to create a custom lifecycle image you can run `make build-image-linux-amd64` (or whatever is the correct os/arch) from the lifecycle repo with your changes.

**Comment by tpodom** [comment]
We ran into this issue as well but it hasn't been a very consistent issue to reproduce. In our scenario, we are running multiple Buildpack packs in parallel in the same source workspace for a monorepo. I'm wondering if there's something about the filesystem changing while the pack is trying to create the tar that trips it up? I'm not sure if maybe some file size change would trigger this error if the file size didn't match the size listed in the header for the entry?

If it is due to the file system changing, I'm not really sure how that could be avoided but maybe there could be more error handling to retry an entry or write out which entry failed? In our scenario if we knew what file was failing it may give a hint on what we could exclude.

**Comment by tpodom** [comment]
I was able to replicate this by having a program start writing to a file and then while that's running I started the Buildpack pack. I believe because the file was changing that the header size didn't match the actual size read causing the error. 

What about detecting the error response "write too long" and then doing something like the following:
1. Check to see if the file size is now different and output a message saying the file changed as it was being read.
2. Have an option to skip changing files to allow the pack to proceed
3. If the file is same size include some generic message indicating the filename with the "write too long" in case there are other cases that can cause it.
